### PR TITLE
[JAX] Support Flax sharding constraints

### PR DESCRIPTION
**Pull request by jberchtold-nvidia**
# Description

Updates our `with_sharding_constraint_by_logical_axes` helper function to support and prefer Flax logical axis rules when they exist in the current context. If no Flax logical axes rules exist in the current context, it will fall back to TE's hardcoded logical axes, though this functionality is no deprecated and will be removed in the future.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

- Update `with_sharding_constraint_by_logical_axes` to check if Flax logical axis rules exist in the current context, and if so call the sharding constraint with `flax.linen.with_logical_constraint`
- Update unfused attention in `transformers.py` to remove duplicate SEQLEN_AXIS usage in a single sharding constraint. This was not flagged before since our TE hardcoded axis system didn't check for this and it was always mapped to `None`/replicated.
- Corrected logical axis rule context setup in encoder examples

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/jax/sharding.py [hunk 1]
```diff
@@ -133,6 +135,19 @@ def with_sharding_constraint_by_logical_axes(
...
+
+        flax_rules = flax.linen.get_logical_axis_rules()
+        if len(flax_rules) > 0:
+            return flax.linen.with_logical_constraint(
+                x, logical_axis_names, fallback=flax.linen.spmd.RulesFallback.NO_CONSTRAINT
+            )
+    except ImportError:
+        pass
+
+    # If no logical axis rules are available from Flax, fallback to TE's hardcoded logical axis rule table
```
Comment (phu0ngng) [inline transformer_engine/jax/sharding.py:160]: Could you raise a warning for soon-to-be deprecated here?

Comment (jberchtold-nvidia) [reply transformer_engine/jax/sharding.py:160]: Good idea, added a warning

File: examples/jax/encoder/test_model_parallel_encoder.py [hunk 2]
```diff
@@ -261,36 +261,37 @@ def train_and_evaluate(args):
...
-            mesh_resource=te.MeshResource(DEVICE_DP_AXIS, DEVICE_TP_AXIS, None, None),
-        ):
+        # Get the base axis rules and extend them with TE's rules.
+        axis_rules = flax.linen.get_logical_axis_rules()
+        axis_rules += ((NAMED_BROADCAST_AXIS, None), (NAMED_TP_AXIS, DEVICE_TP_AXIS))
+        te_extended_axis_rules = te_flax.extend_logical_axis_rules(axis_rules)
+
+        with jax.sharding.Mesh(
+            devices=device_mesh, axis_names=(DEVICE_DP_AXIS, DEVICE_TP_AXIS)
+        ) as mesh, flax.linen.logical_axis_rules(te_extended_axis_rules):
```
Comment (phu0ngng) [inline examples/jax/encoder/test_model_parallel_encoder.py]: Hi, 
I'm curious to learn what the difference is between having the `jax.sharding.Mesh` context before vs after the `te.fp8_autocast` context.

Comment (jberchtold-nvidia) [reply examples/jax/encoder/test_model_parallel_encoder.py]: Afaik it doesn't matter the order of the Mesh vs. fp8_autocast, as long as both are before your model. But the order of `te_flax.extend_logical_axis_rules` and `fp8_autocast` does matter according to the docstring of the former.

So the ordering needs to be:
1. `fp8_autocast` context
2. Create Flax logical axis rule context. To support TE's hardcoded axis system in Flax (which UnfusedDotProductAttention requires), we extend the logical access rules with TE's rule table via `te_flax.extend_logical_axis_rules`, which must be inside an `fp8_autocast`.
3. Create and initialize the model, training loop, etc.

Afaik, the Mesh can come anywhere before item 3.

I just pulled up the `fp8_autocast` to the top and merged the `Mesh` with the `with` block for the logical axis rule context in item 2 to reduce indentation. But if a smaller diff is preferred, I can do Mesh -> fp8_autocast -> logical axis rules

Comment (phu0ngng) [reply examples/jax/encoder/test_model_parallel_encoder.py]: In this case, I would prefer we keep a minimal diff.

Comment (jberchtold-nvidia) [reply examples/jax/encoder/test_model_parallel_encoder.py]: Sounds good, updated to a smaller diff. Mesh -> fp8_autocast order is unchanged. fp8_autocast and logical axis rule order is changed as that was not implemented correctly before this PR.

**Review by phu0ngng — APPROVED** [review]
